FACT-Tools has exit code of 0 despite RuntimeException thrown in Writer

```
$ java -jar target/fact-tools-v1.0.0.jar examples/save_dl1.xml -Dinfile=file:src/main/resources/testDataFile.fits.gz  -Ddrsfile=file:src/main/resources/testDrsFile.drs.fits.gz -Doutfile=file:test.fits
Mar 10 2018 14:48:53 [INFO ] stream.util.PropertiesHandler: Reading properties from URL classpath:/default/settings.properties
Mar 10 2018 14:48:53 [INFO ] stream.util.PropertiesHandler: Reading properties from URL classpath:/default/settings.properties
Mar 10 2018 14:49:10 [ERROR] stream.runtime.AbstractProcess: Error occurred in process '2': Missing keys for processor fact.io.Writer:  photonCharge

Mar 10 2018 14:49:10 [ERROR] stream.runtime.ProcessThread: java.lang.RuntimeException: Missing keys for processor fact.io.Writer:  photonCharge

	at fact.Utils.mapContainsKeys(Utils.java:306)
	at fact.io.Writer.testKeys(Writer.java:67)
	at fact.io.FITSWriter.process(FITSWriter.java:85)
	at stream.runtime.AbstractProcess.process(AbstractProcess.java:126)
	at stream.runtime.AbstractProcess.execute(AbstractProcess.java:184)
	at stream.runtime.ProcessThread.run(ProcessThread.java:129)

Mar 10 2018 14:49:10 [ERROR] stream.runtime.ProcessContainer$3: Process DefaultProcess['2']@c39f790 signaled an error: Missing keys for processor fact.io.Writer:  photonCharge

Exception in thread "Thread-2" java.lang.RuntimeException: Process stopped after exception: Missing keys for processor fact.io.Writer:  photonCharge

	at stream.runtime.ProcessThread.run(ProcessThread.java:170)
(base) maxnoe@maxnoe-mac {master} ~/Uni/FACT/fact-tools
$ echo $?
0

```

Also when Aux data could not be loaded, there is this error:
```
Apr 08 2018 23:55:20 [INFO ] stream.util.PropertiesHandler: Reading properties from URL classpath:/default/settings.properties
Apr 08 2018 23:55:20 [INFO ] stream.util.PropertiesHandler: Reading properties from URL classpath:/default/settings.properties
Apr 08 2018 23:55:21 [WARN ] stream.runtime.setup.ServiceFieldInjection: No service injected for optional service field 'auxService'
Apr 08 2018 23:55:32 [WARN ] fact.features.source.SourcePosition: Using zd and az values from the data item
Apr 08 2018 23:55:41 [WARN ] fact.gainservice.GainService: Time difference to closest gain measurement is more than 5 days
Apr 08 2018 23:55:41 [WARN ] fact.gainservice.GainService: Time difference to closest gain measurement is more than 5 days
Apr 08 2018 23:55:41 [ERROR] fact.auxservice.AuxFileService: Could not load aux file in given directory file:/fact/aux/2011/11/16
Apr 08 2018 23:55:41 [ERROR] fact.Utils: Data does not contain the key shower
Apr 08 2018 23:55:41 [ERROR] stream.runtime.AbstractProcess: Error occurred in process '2': Did not find key 'shower' in the event. For processor:  fact.pixelsets.Length
Apr 08 2018 23:55:41 [ERROR] stream.runtime.ProcessThread: java.lang.RuntimeException: Did not find key 'shower' in the event. For processor:  fact.pixelsets.Length
        at fact.Utils.isKeyValid(Utils.java:329)
        at fact.pixelsets.Length.process(Length.java:28)
        at stream.runtime.AbstractProcess.process(AbstractProcess.java:126)
        at stream.runtime.AbstractProcess.execute(AbstractProcess.java:184)
        at stream.runtime.ProcessThread.run(ProcessThread.java:129)

Apr 08 2018 23:55:41 [ERROR] stream.runtime.ProcessContainer$3: Process DefaultProcess['2']@6bf256fa signaled an error: Did not find key 'shower' in the event. For processor:  fact.pixelsets.Length
Copying /tmp/7879806.1.fact_short/fact_erna_job_id_251320_rgz5q6xz/facttools_output/20111116_016_v1.0.0_std_analysis_level2.fits to /gpfs0/fact/processing/erna/fact-tools/v1.0.0/std_analysis/2011/11/16```

And exit code was still 0.

--- a/pom.xml
+++ b/pom.xml
@@ -4,7 +4,7 @@
   <groupId>de.sfb876</groupId>
   <artifactId>fact-tools</artifactId>
   <name>fact-tools</name>
-  <version>1.0.1</version>
+  <version>1.0.2-SNAPSHOT</version>
   <url>http://sfb876.de/fact-tools/</url>
 
   <description>
@@ -15,7 +15,7 @@
   <properties>
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <streams.scope>compile</streams.scope>
-    <streams.version>1.0.0</streams.version>
+    <streams.version>1.0.5</streams.version>
     <binary.name>fact-tools-${git.commit.id.describe}</binary.name>
     <skipTests>false</skipTests>
     <finalName>fact-tools-${git.commit.id.describe}</finalName>

--- a/src/main/java/fact/datacorrection/InterpolatePixelArray.java
+++ b/src/main/java/fact/datacorrection/InterpolatePixelArray.java
@@ -1,5 +1,6 @@
 package fact.datacorrection;
 
+import fact.Constants;
 import fact.Utils;
 import fact.calibrationservice.CalibrationService;
 import fact.container.PixelSet;
@@ -18,7 +19,10 @@ import java.time.ZonedDateTime;
 /**
  *
  * This Processor interpolates all values for a broken Pixel by the average values of its neighboring Pixels.
-  * @author Kai Bruegge &lt;kai.bruegge@tu-dortmund.de&gt;
+ * It will interpolate the pixels from the GainService.badPixels pixelset as well as all pixels having not-finite
+ * values in the input array.
+ *
+ * @author Kai Bruegge &lt;kai.bruegge@tu-dortmund.de&gt;
  *
  */
 public class InterpolatePixelArray implements Processor {
@@ -48,7 +52,7 @@ public class InterpolatePixelArray implements Processor {
 
         ZonedDateTime timeStamp;
 
-        if (item.containsKey("UnixTimeUTC") == true) {
+        if (item.containsKey("UnixTimeUTC")) {
             Utils.isKeyValid(item, "UnixTimeUTC", int[].class);
             int[] eventTime = (int[]) item.get("UnixTimeUTC");
             timeStamp = Utils.unixTimeUTCToZonedDateTime(eventTime);
@@ -57,7 +61,6 @@ public class InterpolatePixelArray implements Processor {
             // => The 12 bad pixels we have from the beginning on are used.
             timeStamp = ZonedDateTime.of(2000, 1, 1, 0, 0,0,0, ZoneOffset.UTC);
         }
-
         PixelSet badPixelsSet = calibService.getBadPixels(timeStamp);
 
         double[] output;
@@ -73,19 +76,22 @@ public class InterpolatePixelArray implements Processor {
     }
 
     double[] interpolatePixelArray(double[] pixelArray, PixelSet badPixels) {
-        for (CameraPixel pixel: badPixels){
-            CameraPixel[] currentNeighbors = pixelMap.getNeighborsForPixel(pixel);
-            double avg = 0.0;
-            int numNeighbours = 0;
-            for (CameraPixel neighbor: currentNeighbors){
-                if (badPixels.contains(neighbor)) {
-                    continue;
+        for (int chid = 0; chid < Constants.N_PIXELS; chid++){
+            if (badPixels.containsID(chid) || (!Double.isFinite(pixelArray[chid]))) {
+                CameraPixel[] currentNeighbors = pixelMap.getNeighborsFromID(chid);
+
+                double avg = 0.0;
+                int numNeighbours = 0;
+                for (CameraPixel neighbor: currentNeighbors){
+                    if (badPixels.contains(neighbor) || (!Double.isFinite(pixelArray[neighbor.id]))) {
+                        continue;
+                    }
+                    avg += pixelArray[neighbor.id];
+                    numNeighbours++;
                 }
-                avg += pixelArray[neighbor.id];
-                numNeighbours++;
+                checkNumNeighbours(numNeighbours, chid);
+                pixelArray[chid] = avg / numNeighbours;
             }
-            checkNumNeighbours(numNeighbours, pixel.id);
-            pixelArray[pixel.id] = avg / numNeighbours;
         }
         return pixelArray;
     }

--- a/src/main/resources/log4j.properties
+++ b/src/main/resources/log4j.properties
@@ -13,9 +13,10 @@ log4j.appender.A1=org.apache.log4j.ConsoleAppender
 log4j.appender.A1.layout=org.apache.log4j.PatternLayout
 log4j.appender.A1.layout.ConversionPattern=%d{MMM dd yyyy HH:mm:ss} [%-5p] %C: %m%n
 
-log4j.logger.stream=INFO
+log4j.logger.stream=ERROR
+log4j.logger.stream.util=ERROR
 log4j.logger.stream.run=ERROR
-log4j.logger.stream.runtime=INFO
+log4j.logger.stream.runtime=ERROR
 log4j.logger.net.scinotes=INFO
 log4j.logger.net=INFO
 log4j.logger.org.springframework=INFO

--- a/src/test/java/fact/FunctionalTest.java
+++ b/src/test/java/fact/FunctionalTest.java
@@ -18,6 +18,7 @@ import java.util.stream.Collectors;
 
 import static fact.RunFACTTools.runFACTTools;
 import static org.hamcrest.core.Is.is;
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.fail;
 
@@ -58,6 +59,12 @@ public class FunctionalTest {
             String[] args = {f.toAbsolutePath().toString()};
             try {
                 runFACTTools(args);
+            } catch (RunFACTTools.SystemExitException e) {
+                if (e.status != 0) {
+                    log.error("Error executing xml: " + f, e);
+                    failedFilesList.add(f.toString());
+                    counter++;
+                }
             } catch (Exception e) {
                 log.error("Error executing xml: " + f, e);
                 failedFilesList.add(f.toString());
@@ -65,9 +72,9 @@ public class FunctionalTest {
             }
         }
 
-        log.debug("\n\n" + counter + " of " + size + " files in  'examples/studies' failed to execute");
-        log.debug(Arrays.toString(failedFilesList.toArray()));
+        String msg =  counter + " of " + size + " files in  'examples/studies' failed to execute: ";
+        msg += Arrays.toString(failedFilesList.toArray());
 
-        assertThat(failedFilesList.size(), is(0));
+        assertEquals(msg, 0, failedFilesList.size());
     }
 }

--- a/src/test/java/fact/RunFACTTools.java
+++ b/src/test/java/fact/RunFACTTools.java
@@ -1,6 +1,26 @@
 package fact;
 
+import java.security.Permission;
+
 public class RunFACTTools {
+
+    static class SystemExitException extends SecurityException {
+        final public int status;
+
+        SystemExitException(int status) {
+            super("SystemExit called with status code " + status);
+            this.status = status;
+        }
+    }
+
+    static class CatchSystemExitSecurityManager extends SecurityManager {
+        @Override public void checkExit(int status) {
+            throw new SystemExitException(status);
+        }
+
+        @Override public void checkPermission(Permission perm) {}
+    }
+
     public static void runFACTTools(String[] args) throws Exception {
         System.clearProperty("drsfile");
         System.clearProperty("infile");
@@ -10,6 +30,14 @@ public class RunFACTTools {
         System.clearProperty("integralGainFile");
         System.clearProperty("aux_source");
         System.clearProperty("output");
-        fact.run.main(args);
+
+        // Catch the system exit stream calls when an error happens
+        SecurityManager before = System.getSecurityManager();
+        System.setSecurityManager(new CatchSystemExitSecurityManager());
+        try {
+            fact.run.main(args);
+        } finally {
+            System.setSecurityManager(before);
+        }
     }
 }

--- a/src/test/resources/log4j.properties
+++ b/src/test/resources/log4j.properties
@@ -14,6 +14,7 @@ log4j.appender.A1.layout=org.apache.log4j.PatternLayout
 log4j.appender.A1.layout.ConversionPattern=%d{MMM dd yyyy HH:mm:ss} [%-5p] %C: %m%n
 
 log4j.logger.fact=INFO
-#log4j.logger.fact.io=DEBUG
-#log4j.logger.org.apache=DEBUG
-#log4j.logger.com=DEBUG
+log4j.logger.stream=ERROR
+log4j.logger.stream.util=ERROR
+log4j.logger.stream.run=ERROR
+log4j.logger.stream.runtime=ERROR

